### PR TITLE
Move slicetools implementation to cython (PyInf#12655)

### PR DESCRIPTION
**Pull request by ArvidJB**
Constructing `data_dict` is pretty slow:

```
In [2]: with h5py.File('data.h5', 'w') as f:
   ...:     vf = VersionedHDF5File(f)
   ...:     with vf.stage_version('v0') as sv:
   ...:         sv.create_dataset('values', data=np.arange(321 * 54321).reshape((321, 54321)), chunks=(10, 100), maxshape=(None, None))
   ...:
[PYFLYBY] import numpy as np

In [4]: %%pyinstrument
   ...: with h5py.File('data.h5', 'r+') as f:
   ...:     vf = VersionedHDF5File(f)
   ...:     with vf.stage_version('v1') as sv:
   ...:         values = sv['values']
   ...:         values.resize((432, 65432))
   ...:

  _     ._   __/__   _ _  _  _ _/_   Recorded: 11:00:01  Samples:  3915
 /_//_/// /_\ / //_// / //_'/ //     Duration: 6.845     CPU time: 6.262
/   _/                      v4.6.1

Program: /usr/local/bin/py

6.844 <module>  <ipython-input-4-25703e99d0a0>:1
|- 3.034 InMemoryDataset.resize  versioned_hdf5/wrappers.py:609
|  |- 1.292 InMemoryDatasetID.data_dict  versioned_hdf5/wrappers.py:1298
|  |  |- 0.690 <dictcomp>  versioned_hdf5/wrappers.py:1327
|  |  |  |- 0.600 spaceid_to_slice  versioned_hdf5/slicetools.py:6
|  |  |  |  `- 0.506 [self]  versioned_hdf5/slicetools.py
|  |  |  `- 0.090 [self]  versioned_hdf5/wrappers.py
|  |  |- 0.317 <listcomp>  versioned_hdf5/wrappers.py:1317
|  |  |  `- 0.301 [self]  versioned_hdf5/wrappers.py
|  |  `- 0.211 [self]  versioned_hdf5/wrappers.py
|  |- 0.951 InMemoryDataset.get_index  versioned_hdf5/wrappers.py:655
|  |  `- 0.951 InMemoryDataset.__getitem__  h5py/_hl/dataset.py:749
|  |     `- 0.949 Reader.read  <built-in>
|  `- 0.677 [self]  versioned_hdf5/wrappers.py
|- 2.800 _GeneratorContextManager.__exit__  contextlib.py:141
|  `- 2.800 VersionedHDF5File.stage_version  versioned_hdf5/api.py:267
...
|- 0.762 _GeneratorContextManager.__enter__  contextlib.py:132
|  `- 0.762 VersionedHDF5File.stage_version  versioned_hdf5/api.py:267
...
`- 0.234 File.__exit__  h5py/_hl/files.py:601
   `- 0.234 File.close  h5py/_hl/files.py:576
```

Most of this is simple overhead. By re-implementing the building of the data_dict in Cython we can speed up the performance 40x-50x:

```
|- 0.509 InMemoryDataset.resize  versioned_hdf5/wrappers.py:608
|  |- 0.265 InMemoryDataset.get_index  versioned_hdf5/wrappers.py:654
|  |  `- 0.264 InMemoryDataset.__getitem__  h5py/_hl/dataset.py:749
|  |     `- 0.264 Reader.read  <built-in>
|  |- 0.154 [self]  versioned_hdf5/wrappers.py
|  `- 0.030 InMemoryDatasetID.data_dict  versioned_hdf5/wrappers.py:1297
```

### CODE REVIEW

File: versioned_hdf5/slicetools.pyx [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+cdef _spaceid_to_slice(space_id: hid_t):
+    sel_type: H5S_sel_type = H5Sget_select_type(space_id)
+
+    if sel_type == H5S_sel_type.H5S_SEL_ALL:
+        return Tuple()
+    elif sel_type == H5S_sel_type.H5S_SEL_HYPERSLABS:
+        slices: list = []
+
+        rank: cython.int = H5Sget_simple_extent_ndims(space_id)
```
Comment (peytondmurray) [inline versioned_hdf5/slicetools.pyx:84]: If `malloc` fails do we need to raise an exception?

File: versioned_hdf5/slicetools.pyx [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+            try:
+                dataset_buf_len: ssize_t = strlen(raw_data_str) + 1
+                dataset_buf: cython.p_char = <char *>malloc(dataset_buf_len)
+
+                try:
+                    for j in range(virtual_count):
+                        H5Pget_virtual_filename(dcpl_id, j, filename_buf, filename_buf_len)
+                        if strncmp(filename_buf, ".", filename_buf_len) != 0:
+                            raise ValueError('Virtual dataset filename mismatch, expected "."')
```
Comment (peytondmurray) [inline versioned_hdf5/slicetools.pyx:160]: :+1: Good idea to check this, definitely one of the weirder features of h5py.

File: versioned_hdf5/slicetools.pyx [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+            raw_data_name_bytes: bytes = raw_data_name.encode('utf8')
+            # this a reference to the internal buffer of raw_data_name, do not free!
+            raw_data_str: cython.p_char = raw_data_name_bytes
+
+            filename_buf_len: ssize_t = 2
+            filename_buf: cython.p_char = <char *>malloc(filename_buf_len)
+
+            try:
+                dataset_buf_len: ssize_t = strlen(raw_data_str) + 1
+                dataset_buf: cython.p_char = <char *>malloc(dataset_buf_len)
```
Comment (peytondmurray) [inline versioned_hdf5/slicetools.pyx:154]: Same question about malloc as above :point_up: - do we need to raise an exception if this fails?

**Review by peytondmurray — COMMENTED** [review]
Thanks for this! A few small comments, this is close to ready. I guess we don't need additional tests, because if we weren't able to create `data_dict` objects we'd presumably be unable to create datasets.

File: versioned_hdf5/slicetools.pyx [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    cdef htri_t H5Sget_regular_hyperslab(hid_t spaceid, hsize_t* start, hsize_t* stride, hsize_t* count, hsize_t* block) except <htri_t>-1
+
+def spaceid_to_slice(space) -> Tuple:
+    """
+    Convert an h5py spaceid object into an ndindex index
+
+    The resulting index is always a Tuple index.
+    """
+
+    from h5py import h5s
```
[the remark this answers is not in the document]
Comment (ArvidJB) [reply versioned_hdf5/slicetools.pyx:47]: Good point, moving it to the top.

**Pull request merged**